### DISCUSSION TITLE
How to load heterogeneous temporal graph with the combination of  HeteroData and TemporalData

### DISCUSSION THREAD

**Discussion by rangerhu** [pyg-team/pytorch_geometric#4658] [Q&A] [answered]
In the case of https://pytorch-geometric.readthedocs.io/en/latest/notes/load_csv.html, if I want to combine the timestamp attr from rating.csv, thus making the constructing graph as a heterogeneous temporal graph, How shall I change the code in https://github.com/pyg-team/pytorch_geometric/blob/master/examples/hetero/load_csv.py to reach this goal.
PS: I noticed that those data types under TORCH_GEOMETRIC.DATA class include HeteroData() and TemporalData() which are designed for Heterogeneous graph and Temporal graph respectively, if I want to construct heterogeneous temporal graph, then how can I combine HeteroData() with and TemporalData() to load my data? 
Can't be appreciated it too much if you can find time to answer my question.

**Comment by rusty1s** [comment, marked as answer]
I think in these scenarios that `HeteroData` is still the preferred choice here. Note that we are currently working on providing better support for this. For example, we just added support for this in `torch-sparse` (requires manual installation from source for now), and with the node-level `time_attr` attribute of the `NeighborLoader` class. With this, the `NeighborLoader` class performs temporal sampling that avoids leaking future information.

**Reply by rangerhu** [reply]
Thank you for your reply, sir. Is that mean in order to reach my goal, I will need to do the following work?
1.  assign the timestamp value to `data['user', 'rates', 'movie'].time_attr` (in the case of [Pyg docs MovieLens dataset Case](https://pytorch-geometric.readthedocs.io/en/latest/notes/load_csv.html#))
2.  upgrade my pytorch-sparse library to the latest version by manualling installation from source
3.  Convert an existing temporal GNN(e.g. TGN) to heterogeneous temporal GNN using `torch_geometric.nn.to_hetero()` to embedding the heterogeneous temporal graph.
PS: my current version of relevant dependency:
pytorch                        1.10.1
pytorch-sparse            0.6.13
pyg                              2.0.4

**Reply by rangerhu** [reply]
From the GitHub page of [pytorch-sparse](https://github.com/rusty1s/pytorch_sparse/releases/tag/0.6.13) , it seems the latest version of the library is 0.6.13, which means I will don't have to upgrade my pytorch-sparse library?

**Reply by rusty1s** [reply]
1. Currently, `time_attr` needs to be a node-level attribute. This is a current limitation, sorry about that :(
2. Temporal sampling support is not yet released. It will get integrated in a new `torch-sparse` release. For now, cloning the repo and running `pip install -e .` should work (but may take some time to compile).
3. Yes. You can also compare your model to TGN. This would be a very interesting case study.

**Reply by rangerhu** [reply]
got it, thank you,sir ~

**Reply by shairal1** [reply]
Hey , I  also want something like this . Would really like to know if you figured out a way to do this :) .Thanks @rangerhu
